File: programmers/level2/할인 행사.py
from collections import defaultdict


def check(wantDict, discountDict):
    count = 0  # 제품 확인
    for key, value in wantDict.items():
        if discountDict[key] != 0:  # 해당 키 값이 존재한다면
            if discountDict[key] == value:  # 키값과 동일한지 체크
                count += 1
            else:  # 키값 존재하지 않으면 실패
                return False
        else:  # 해당 키 값 존재하지 않으면 실패
            return False

    if len(discountDict) != count:  # 키값의 개수와 동일하지 않으면 False
        return False
    else:  # 동일하면 True
        return True


def solution(want, number, discount):
    answer = 0
    wantDict = defaultdict(int)

    for i in range(len(want)):  # 원하는 제품 체크
        wantDict[want[i]] = number[i]

    for i in range(len(discount)):
        discountDict = defaultdict(int)  # 10일동안 구매할 수 있는 제품 딕셔너리
        if i + 10 <= len(discount):  # 길이에 따라 10일동안 구매할 수 있는 제품 딕셔너리 추가
            for j in range(i, i + 10):
                discountDict[discount[j]] += 1
        else:
            for j in range(i, len(discount)):
                discountDict[discount[j]] += 1

        if check(wantDict, discountDict):  # 함수로 체크
            answer += 1  # 날짜 플러스

    return answer

File: programmers/level2/할인 행사.java
import java.util.*;

class Solution {
    public int solution(String[] want, int[] number, String[] discount) {
        int answer = 0;
        Map<String, Integer> wantMap = new HashMap<>();

        for (int i = 0; i < want.length; i++) {
            wantMap.put(want[i], number[i]);
        }

        for(int i = 0; i < discount.length; i++) {
            Map<String, Integer> discountMap = new HashMap<>();
            int end = i + 10 > discount.length ? discount.length : i + 10;
            for (int j = i; j < end; j++) {
                discountMap.put(discount[j], discountMap.getOrDefault(discount[j], 0) + 1);
            }
            if (check(wantMap, discountMap)) {
                answer += 1;
            }
        }
        return answer;
    }

    public boolean check(Map<String, Integer> wantMap, Map<String, Integer> discountMap) {
        for (Map.Entry<String, Integer> want: wantMap.entrySet()) {
            if (discountMap.containsKey(want.getKey())) {
                if (want.getValue() - discountMap.get(want.getKey()) > 0) {
                    return false;
                }
            } else {
                return false;
            }
        }
        return true;
    }
}